File: mycuration/src/test/java/com/phicdy/mycuration/presenter/FeedUrlHookPresenterTest.java
package com.phicdy.mycuration.presenter;

import android.content.Context;
import android.content.Intent;

import com.phicdy.mycuration.db.DatabaseAdapter;
import com.phicdy.mycuration.rss.Feed;
import com.phicdy.mycuration.rss.RssParseExecutor;
import com.phicdy.mycuration.rss.RssParseResult;
import com.phicdy.mycuration.rss.RssParser;
import com.phicdy.mycuration.rss.UnreadCountManager;
import com.phicdy.mycuration.task.NetworkTaskManager;
import com.phicdy.mycuration.view.FeedUrlHookView;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;

import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;

public class FeedUrlHookPresenterTest {

    private NetworkTaskManager networkTaskManager;
    private DatabaseAdapter adapter;
    private UnreadCountManager unreadCountManager;
    private FeedUrlHookPresenter presenter;
    private RssParser parser;

    @Before
    public void setup() {
        NetworkTaskManager.INSTANCE.setUp(Mockito.mock(Context.class));
        networkTaskManager = NetworkTaskManager.INSTANCE;
        adapter = Mockito.mock(DatabaseAdapter.class);
        unreadCountManager = Mockito.mock(UnreadCountManager.class);
        parser = Mockito.mock(RssParser.class);
        presenter = new FeedUrlHookPresenter(
                adapter, unreadCountManager, networkTaskManager, parser);

    }

    @Test
    public void testOnCreate() {
        // For coverage
        presenter.setView(new MockView());
        presenter.create();
        assertTrue(true);
    }

    @Test
    public void testOnResume() {
        // For coverage
        presenter.setView(new MockView());
        presenter.create();
        presenter.resume();
        assertTrue(true);
    }

    @Test
    public void receiverIsUnregisteredInAfterPause() {
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.pause();
        assertFalse(view.isReceiverRegistered);
    }

    @Test
    public void finishWhenInvalidActionComes() {
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.handle("hogehoge", "http://www.google.com");
        assertTrue(view.isFinished);
    }

    @Test
    public void finishWhenEmptyActionComes() {
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.handle("", "http://www.google.com");
        assertTrue(view.isFinished);
    }

    @Test
    public void registerReceiverWhenActionViewAndUrlComes() {
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.handle(Intent.ACTION_VIEW, "http://www.google.com");
        assertTrue(view.isReceiverRegistered);
    }

    @Test
    public void toastShowsWhenActionViewAndInvalidUrlComes() {
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.handle(Intent.ACTION_VIEW, "hogehoge");
        assertTrue(view.isInvalidUrlErrorToastShowed);
    }

    @Test
    public void registerReceiverWhenActionSendAndUrlComes() {
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.handle(Intent.ACTION_SEND, "http://www.google.com");
        assertTrue(view.isReceiverRegistered);
    }

    @Test
    public void toastShowsWhenActionSendAndInvalidUrlComes() {
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.handle(Intent.ACTION_SEND, "hogehoge");
        assertTrue(view.isInvalidUrlErrorToastShowed);
    }

    @Test
    public void viewDoesNotFinishWhenInvalidActionComes() {
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.handleFinish("hogehoge", "http://www.google.com", NetworkTaskManager.REASON_NOT_FOUND);
        assertFalse(view.isFinished);
    }

    @Test
    public void viewFinishesWhenFinishAddFeedActionComes() {
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.callback.failed(RssParseResult.NOT_FOUND);
        assertTrue(view.isFinished);
    }

    @Test
    public void toastShowsWhenFinishAddFeedActionComesWithNotRssHtmlError() {
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.callback.failed(RssParseResult.NON_RSS_HTML);
        assertTrue(view.isGenericErrorToastShowed);
    }

    @Test
    public void toastShowsWhenFinishAddFeedActionComesWithInvalidUrlError() {
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.callback.failed(RssParseResult.INVALID_URL);
        assertTrue(view.isInvalidUrlErrorToastShowed);
    }

    private class MockView implements FeedUrlHookView {
        private boolean isReceiverRegistered = false;
        private boolean isSuccessToastShowed = false;
        private boolean isInvalidUrlErrorToastShowed = false;
        private boolean isGenericErrorToastShowed = false;
        private boolean isFinished = false;

        @Override
        public void registerFinishAddReceiver() {
            isReceiverRegistered = true;
        }

        @Override
        public void unregisterFinishAddReceiver() {
            isReceiverRegistered = false;
        }

        @Override
        public void showSuccessToast() {
            isSuccessToastShowed = true;
        }

        @Override
        public void showInvalidUrlErrorToast() {
            isInvalidUrlErrorToastShowed = true;
        }

        @Override
        public void showGenericErrorToast() {
            isGenericErrorToastShowed = true;
        }

        @Override
        public void finishView() {
            isFinished = true;
        }
    }
}

File: feature_feed_url_hook/src/test/java/com/phicdy/mycuration/feedurlhook/FeedUrlHookPresenterTest.kt
package com.phicdy.mycuration.feedurlhook

import android.content.Intent
import com.phicdy.mycuration.data.repository.RssRepository
import com.phicdy.mycuration.domain.rss.RssParseResult
import com.phicdy.mycuration.domain.rss.RssParser
import com.phicdy.mycuration.domain.rss.RssUrlHookIntentData
import com.phicdy.mycuration.domain.task.NetworkTaskManager
import kotlinx.coroutines.runBlocking
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.times
import org.mockito.kotlin.verify

class FeedUrlHookPresenterTest {

    private val networkTaskManager = mock<NetworkTaskManager>()
    private val rssRepository = mock<RssRepository>()
    private lateinit var parser: RssParser
    private lateinit var view: FeedUrlHookView

    @Before
    fun setup() {
        parser = mock()
        view = mock()
    }

    @Test
    fun `when empty action then finish`() = runBlocking {
        val presenter = FeedUrlHookPresenter(view, RssUrlHookIntentData("", "", ""), rssRepository,
                networkTaskManager, this, parser)
        presenter.create()
        verify(view, times(1)).finishView()
    }

    @Test
    fun `when invalid action comes then finish`() = runBlocking {
        val presenter = FeedUrlHookPresenter(view, RssUrlHookIntentData("hogehoge", "http://www.google.com", ""),
                rssRepository, networkTaskManager, this, parser)
        presenter.create()
        verify(view, times(1)).finishView()
    }

    @Test
    fun `when action view and invalid url comes then toast shows`() = runBlocking {
        val presenter = FeedUrlHookPresenter(view, RssUrlHookIntentData(Intent.ACTION_VIEW, "hogehoge", ""),
                rssRepository, networkTaskManager, this, parser)
        presenter.create()
        verify(view, times(1)).showInvalidUrlErrorToast()
    }

    @Test
    fun `when action send and invalid url comes then toast shows`() = runBlocking {
        val presenter = FeedUrlHookPresenter(view, RssUrlHookIntentData(Intent.ACTION_SEND, "", "hogehoge"),
                rssRepository, networkTaskManager, this, parser)
        presenter.create()
        verify(view, times(1)).showInvalidUrlErrorToast()
    }

    @Test
    fun `when finish add feed action comes then view finishes`() = runBlocking {
        val presenter = FeedUrlHookPresenter(view, RssUrlHookIntentData("", "", ""), rssRepository,
                networkTaskManager, this, parser)
        presenter.callback.failed(RssParseResult.FailedReason.NOT_FOUND, "")
        verify(view, times(1)).finishView()
    }

    @Test
    fun `when finish add feed action comes with not RSS html error then toast shows`() = runBlocking {
        val presenter = FeedUrlHookPresenter(view, RssUrlHookIntentData("", "", ""), rssRepository,
                networkTaskManager, this, parser)
        presenter.create()
        presenter.callback.failed(RssParseResult.FailedReason.NON_RSS_HTML, "")
        verify(view, times(1)).showGenericErrorToast()
    }

    @Test
    fun `when finish add feed action comes with invalid url error then toast shows`() = runBlocking {
        val presenter = FeedUrlHookPresenter(view, RssUrlHookIntentData("", "", ""), rssRepository,
                networkTaskManager, this, parser)
        presenter.create()
        presenter.callback.failed(RssParseResult.FailedReason.INVALID_URL, "")
        verify(view, times(1)).showInvalidUrlErrorToast()
    }

    @Test
    fun `when failed then the url will be tracked`() = runBlocking {
        val presenter = FeedUrlHookPresenter(view, RssUrlHookIntentData("", "", ""), rssRepository,
                networkTaskManager, this, parser)
        presenter.create()
        val failUrl = "http://www.google.com"
        presenter.callback.failed(RssParseResult.FailedReason.INVALID_URL, failUrl)
        verify(view, times(1)).trackFailedUrl(failUrl)
    }
}
